### PR TITLE
Add maven profile for -Xlint:deprecation compiler option

### PR DESCRIPTION
**Pull request by leveyja**
## Description
Add maven profile for -Xlint:deprecation compiler option.

Compiling with ./mvnw [...] -Pdeprecations will warn about deprecated method and class usages.

This PR includes individual commits for resolution of deprecated usages of 3rd party libraries.

> Is this change a fix, improvement, new feature, refactoring, or other?

Refactoring

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Removal of deprecated method usages across all components

> How would you describe this change to a non-technical end user or system administrator?

Maintenance of 3rd party libraries to ensure our usages of them are non-deprecated.

## Documentation

(:white_check_mark:) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(:white_check_mark:) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: client/trino-client/src/main/java/io/trino/client/JsonCodec.java [hunk 1]
```diff
@@ -30,7 +31,8 @@
 public class JsonCodec<T>
 {
     // copy of https://github.com/airlift/airlift/blob/master/json/src/main/java/io/airlift/json/ObjectMapperProvider.java
-    static final Supplier<ObjectMapper> OBJECT_MAPPER_SUPPLIER = () -> new ObjectMapper()
+    // TODO: fix deprecated usages in Airlift's ObjectMapperProvider
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/JsonCodec.java:34]: don't add TODO here, it belongs to Airlift

**Comment by findepi** [comment]
I like the idea of addressing deprecation warnings. In fact, it would be nice to force zero deprecation policy (either address or suppress, but don't ignore).

From review perspective, this PR needs to be split into multiple smaller ones -- I am happy to approve and merge some of these changes, and don't have enough context for some others.

File: plugin/trino-bigquery/pom.xml [hunk 2]
```diff
@@ -179,17 +179,6 @@
             <artifactId>guava</artifactId>
         </dependency>
 
-        <dependency>
-            <groupId>com.google.http-client</groupId>
```
[the remark this answers is not in the document]
Comment (leveyja) [reply plugin/trino-bigquery/pom.xml:183]: 👍 I'll prepare separate PRs for each category of deprecation.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystem.java [hunk 3]
```diff
@@ -489,7 +489,7 @@ public void testGetMetadataNotFound()
         }
     }
 
-    @Test
+    @Test @SuppressWarnings("deprecation") // TODO: upgrade to AmazonS3EncryptionClientV2
```
[the remark this answers is not in the document]
Comment (leveyja) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystem.java:492]: Right - I attempted to upgrade to `AmazonS3EncryptionClientV2` but it's not a drop-in replacement. Left the "TODO" here to note that this requires work (not just SuppressWarnings and walk away). I'll create Issues for all TODOs that are added next to SuppressWarnings.

File: core/trino-main/src/main/java/io/trino/operator/scalar/JsonFunctions.java [hunk 4]
```diff
@@ -58,8 +58,7 @@
             .disable(CANONICALIZE_FIELD_NAMES)
             .build();
 
-    private static final JsonFactory MAPPING_JSON_FACTORY = new MappingJsonFactory()
-            .disable(CANONICALIZE_FIELD_NAMES);
```
[the remark this answers is not in the document]
Comment (leveyja) [reply core/trino-main/src/main/java/io/trino/operator/scalar/JsonFunctions.java:62]: It's still there (see line below), just the only way to remove the `.disable(Feature)` usage is to pass the pre-configured JSON_FACTORY to `new MappingJsonFactory(JSON_FACTORY, null)` (`null` causes the ObjectMapper/codec to be created using the supplsed JSON_FACTORY).

**Comment by leveyja** [comment]
Thanks for the feedback @findepi! I will close this + open individual PRs for each category of deprecation fix.

**Comment by leveyja** [comment]
`-Xlint:deprecation` change here: https://github.com/trinodb/trino/pull/12628
`Resolve all com.google.common.io.Files deprecation warnings` here: https://github.com/trinodb/trino/pull/12629

**Pull request closed without merging**